### PR TITLE
Added new file upload preview

### PR DESCRIPTION
**Pull request by Buildstarted**
- Generic callback handler for uploads to allow different upload types
- Image preview when uploading an image
- Filename preview when uploading a non-image
- Cancel button added to upload

### CODE REVIEW

**Comment by samandmoore** [comment]
pics or it didn't happen

**Comment by Buildstarted** [comment]
pics were posted in jabbr. :)

### uploading an image



### uploading a generic file

File: JabbR/Chat.ui.js [hunk 1]
```diff
@@ -1166,66 +1169,113 @@
...
 
-                            $hiddenFile.val(''); //hide upload dialog
-                        }
-                    };
+                });
+            });
 
-                ui.addMessage('Uploading \'' + name + '\'.', 'broadcast');
+            if (window.File && window.FileReader && window.FileList && window.Blob) {
+                console.log("file reader support");
```
Comment (davidfowl) [inline JabbR/Chat.ui.js]: console.log doesn't work if the console isn't visible in IE.

**Comment by samandmoore** [comment]
can it say something like "upload is a document, no preview available" instead of just being blank for documents?

**Comment by Buildstarted** [comment]
@alfhenrik was going to add some custom icons to display in lieu of text but that's a good addition

**Comment by samandmoore** [comment]
@Buildstarted that was me, not @alfhenrik :trollface:

**Comment by Buildstarted** [comment]
no no, i was telling you who was going to - not replying to him :)

File: JabbR/Chat.ui.js [hunk 2]
```diff
@@ -1166,66 +1169,108 @@
...
+            $.imagePaste(function (file) {
+                ui.showClipboardUpload(file, "clipboard", function() {
+                    var name = "clipboard-data",
+                        uploader = {
+                            submitFile: function (connectionId, room) {
+                                $fileConnectionId.val(connectionId);
+
+                                $fileRoom.val(room);
+
+                                //$uploadForm.submit();
```
Comment (davidfowl) [inline JabbR/Chat.ui.js]: Remove the dead code.

File: JabbR/Chat.ui.js [hunk 3]
```diff
@@ -1166,66 +1169,108 @@
...
             });
 
-            $hiddenFile.change(function () {
-                if (!$hiddenFile.val()) {
+            $hiddenFile.change(function (e) {
+                var file = e.target.files[0];
+                
+                if (!file.type.match('image.*')) {
+                    //it's a standard file - just upload it for now
+                    ui.showClipboardUpload({ data: e.target, name: file.name }, "non-image", function() {
```
Comment (davidfowl) [inline JabbR/Chat.ui.js]: Can we change the function name?

File: JabbR/Chat.ui.js [hunk 4]
```diff
@@ -1166,66 +1169,108 @@
...
+                        var path = $hiddenFile.val(),
+                            slash = path.lastIndexOf('\\'),
+                            name = path.substring(slash + 1),
+                            uploader = {
```
Comment (davidfowl) [inline JabbR/Chat.ui.js]: This seems to be in 2 places. Why?

File: JabbR/Chat.ui.js (hunk 3 above)
Comment (Buildstarted) [inline JabbR/Chat.ui.js]: Changed to `showUploadPreview`

File: JabbR/Chat.ui.js (hunk 4 above)
Comment (Buildstarted) [inline JabbR/Chat.ui.js]: Refactored out to separate function call

File: JabbR/Chat.ui.js [hunk 5]
```diff
@@ -1982,9 +2010,26 @@
             });
             $helpPopup.modal();
         },
-        showClipboardUpload: function (file) {
+        showUploadPreview: function (file, type, uploader) {
```
Comment (davidfowl) [inline JabbR/Chat.ui.js]: Is this function used outside of this file? If no then put it with the other privates at the top of the file.

File: JabbR/Chat.ui.js [hunk 6]
```diff
@@ -1166,66 +1171,89 @@
...
                 $clipboardUpload.modal('hide');
             });
 
-            $.imagePaste(function (file) {
-                ui.showClipboardUpload(file);
+            $previewCancelButton.on("click", function() {
+                $clipboardUpload.modal('hide');
             });
+            
+            function initializeUploadPreview(e, file, type) {
```
Comment (davidfowl) [inline JabbR/Chat.ui.js]: Move this function with the other private ones at the top of the file.

File: JabbR/Views/Home/index.cshtml [hunk 7]
```diff
@@ -424,18 +424,26 @@
             <div id="loadingRoomHistory" class="loading-room-history alert alert-info">
                 <i class="icon-spinner icon-spin"></i> Loading previous messages...
             </div>
-            <div id="jabbr-clipboard-upload" class="modal hide fade">
+            <div id="connectionStatus" class="connection-status" data-toggle="popover" data-placement="left">
```
Comment (davidfowl) [inline JabbR/Views/Home/index.cshtml:427]: What happend here? Why does this all look new?

File: JabbR/Chat.ui.js [hunk 8]
```diff
@@ -713,6 +716,52 @@
...
         $lobbyRoomFilterForm.submit();
     }
+    
+    function showUploadPreview(file, type, uploader) {
+        $uploadCallback = uploader;
+        $imageUploadPreview.show();
+        $unknownUploadPreview.hide();
+        //set image url
+        if (file.dataURL) {
+            $clipboardUpload.find("h3").text("Uploading from clipboard");
```
Comment (davidfowl) [inline JabbR/Chat.ui.js]: We use single quotes

File: JabbR/Chat.ui.js [hunk 9]
```diff
@@ -1258,66 +1309,66 @@
...
-                            }).done(function (result) {
-                                //remove image from preview
-                                $clipboardUploadPreview.attr("src", "");
-                            });
-
-                            $hiddenFile.val(''); //hide upload dialog
-                        }
-                    };
+            $.imagePaste(function (file) {
+                showUploadPreview(file, "clipboard", function() {
```
Comment (davidfowl) [inline JabbR/Chat.ui.js]: Single quotes

**Pull request merged**